### ISSUE TITLE
Script 'scripts/collapse_isoforms_by_sam.py' not found in metadata

### ISSUE THREAD

**Issue by jdmontenegro** [Magdoll/cDNA_Cupcake#140]
Hi,
I am trying to install cupcake from the git repo and I am doing it in the following way:

```
conda create --name cupcake python=3.7 anaconda
conda activate cupcake
pip install bcbio-gff pysam sklearn biopython bx-python
git clone https://github.com/Magdoll/cDNA_Cupcake.git
cd cDNA_Cupcake.git
python setup.py build
python setup.py install
> which collapse_isoforms_by_sam.py
~/.conda/envs/cupcake/bin/collapse_isoforms_by_sam.py
> collapse_isoforms_by_sam.py --help
Traceback (most recent call last):
  File "/home/user/jmontenegro/.conda/envs/cupcake/bin/collapse_isoforms_by_sam.py", line 4, in <module>
    __import__('pkg_resources').run_script('cupcake==15.1.0', 'collapse_isoforms_by_sam.py')
  File "/home/user/jmontenegro/.conda/envs/cupcake/lib/python3.7/site-packages/pkg_resources/__init__.py", line 665, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/user/jmontenegro/.conda/envs/cupcake/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1453, in run_script
    .format(**locals()),
pkg_resources.ResolutionError: Script 'scripts/collapse_isoforms_by_sam.py' not found in metadata at '/proj/jmontenegro/software/cDNA_Cupcake/cupcake.egg-info'

```
I found that installation formula after playing with several different combinations of installation commands. This formula is the only one that does not produce any errors and yet I get this error after installing the cupcake git repo.

Any suggestions how to get around this issue would be more than welcome.

Regards,

Juan D. Montenegro

**Comment by moschmi** [comment]
Same problem here. I wanted to update from Cupcake 12.2, Created a new env etc... But always get the same error like @jdmontenegro 

Marcel Schmidt

**Comment by Magdoll** [comment]
Hi @jdmontenegro @moschmi 
Did you do the which in a different directory other than Cupcake? like go to home directory or something first?

```
cd ~/
which collapse_isoforms_by_sam.py
```

What does this do?
-Liz

**Comment by moschmi** [comment]
then I get this as well:
/etc/samba/mos6/user_homes/mos6/anaconda3/bin/collapse_isoforms_by_sam.py

**Comment by Magdoll** [comment]
@moschmi  right but do you get the same error if you do `collapse_isoforms_by_sam.py --help`?

**Comment by moschmi** [comment]
collapse_isoforms_by_sam.py 
Traceback (most recent call last):
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/bin/collapse_isoforms_by_sam.py", line 4, in <module>
    __import__('pkg_resources').run_script('cupcake==16.0.0', 'collapse_isoforms_by_sam.py')
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 665, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1453, in run_script
    .format(**locals()),
pkg_resources.ResolutionError: Script 'scripts/collapse_isoforms_by_sam.py' not found in metadata at '/etc/samba/mos6/user_homes/mos6/cDNA_Cupcake/cupcake.egg-info'

**Comment by moschmi** [comment]
Ok, I'll delete '/etc/samba/mos6/user_homes/mos6/cDNA_Cupcake/cupcake.egg-info' and reinstall

**Comment by moschmi** [comment]
same error

**Comment by moschmi** [comment]
@Magdoll 
I tried to run this also on other computer/servers. Even with old version 12.2 gives me the same error.

**Comment by moschmi** [comment]
My Mac has this error:
collapse_isoforms_by_sam.py --help
Traceback (most recent call last):
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (six 1.12.0 (/Users/mos61/opt/anaconda3/lib/python3.7/site-packages), Requirement.parse('six>=1.13.0'), {'bx-python'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mos61/opt/anaconda3/bin/collapse_isoforms_by_sam.py", line 4, in <module>
    __import__('pkg_resources').run_script('cupcake==16.0.0', 'collapse_isoforms_by_sam.py')
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3250, in <module>
    @_call_aside
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3234, in _call_aside
    f(*args, **kwargs)
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3263, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 585, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 598, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/Users/mos61/opt/anaconda3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (six 1.12.0 (/Users/mos61/opt/anaconda3/lib/python3.7/site-packages), Requirement.parse('six>=1.13.0'), {'bx-python'})

**Comment by moschmi** [comment]
linux server has this error:
collapse_isoforms_by_sam.py --input isoseq_output/Lneg_7_1.clust.hq.fasta.gz -s mmp_output/Lneg_7_1_hq_mmp.sort.sam  -o  Lneg_7_1_hq_mmp_
Traceback (most recent call last):
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/bin/collapse_isoforms_by_sam.py", line 4, in <module>
    __import__('pkg_resources').run_script('cupcake==16.0.0', 'collapse_isoforms_by_sam.py')
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 665, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/pkg_resources/__init__.py", line 1463, in run_script
    exec(code, namespace, namespace)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/cupcake-16.0.0-py3.7-linux-x86_64.egg/EGG-INFO/scripts/collapse_isoforms_by_sam.py", line 245, in <module>
    main(args)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/cupcake-16.0.0-py3.7-linux-x86_64.egg/EGG-INFO/scripts/collapse_isoforms_by_sam.py", line 169, in main
    check_ids_unique(args.input, is_fq=args.fq)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/cupcake-16.0.0-py3.7-linux-x86_64.egg/cupcake/tofu/utils.py", line 11, in check_ids_unique
    for r in SeqIO.parse(open(fa_or_fq_filename), 'fastq' if is_fq else 'fasta'):
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/biopython-1.78-py3.7-linux-x86_64.egg/Bio/SeqIO/Interfaces.py", line 73, in __next__
    return next(self.records)
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/biopython-1.78-py3.7-linux-x86_64.egg/Bio/SeqIO/FastaIO.py", line 198, in iterate
    for title, sequence in SimpleFastaParser(handle):
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/biopython-1.78-py3.7-linux-x86_64.egg/Bio/SeqIO/FastaIO.py", line 47, in SimpleFastaParser
    for line in handle:
  File "/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)

**Comment by Magdoll** [comment]
Hi @moschmi ,
I'm not sure what to make of this. It just looks like the library imports are messed up :-( 

Besides recommending making sure all your old cupcake eggs are deleted 
(ex: delete everything `/etc/samba/mos6/user_homes/mos6/anaconda3/envs/anaCogent3/lib/python3.7/site-packages/cupcake*) and doing a clean Cupcake install, not sure what else can be done.

Or I guess make a new conda env...

**Comment by moschmi** [comment]
@Magdoll 
That's what I did. I made a new conda env and installed cupcake on top of that. I'll try it again on another clean computer.

**Comment by moschmi** [comment]
It seems to work if I omit PYTHONPATH below
export PYTHONPATH=~/cDNA_Cupcake/sequence/:$PYTHONPATH
export PYTHONPATH=~/cDNA_Cupcake/:$PYTHONPATH

**Comment by Magdoll** [comment]
Hi @moschmi ,
Oh yeeah that is the reason why. You probably should be doing this instead:

```
export PATH=$PATH:<path_to_cDNA_Cupcake>/sequence
```

You want to make sure $PATH comes first. Also, it is `<path_to_cDNA_Cupcake>/sequence` that contains the scripts you may want to call directly.

**Comment by mmarchin** [comment]
This is still happening for me and I can't tell why.

> It seems to work if I omit PYTHONPATH below
> export PYTHONPATH=~/cDNA_Cupcake/sequence/:$PYTHONPATH export PYTHONPATH=~/cDNA_Cupcake/:$PYTHONPATH

THANK YOU! After fiddling with this for far too long, keeping the cDNA_Cupcake out of my PYTHONPATH worked for me.

**Comment by LilyLuyang** [comment]
Hi everyone,

I also encountered this same error when in the conda subenv, and I also tried aforementioned solutions but still unsuccessful. Therefore, I just recreate conda environment, while another question occurred: 'pkg_resources.DistributionNotFound: The 'sklearn' distribution was not found and is required by cupcake', so I tried several ways to address it. Finally, it works after changing 'sklearn' into 'scikit-learn' on line 40 in setup.py, and 'python setup.py build & python setup.py install' to configure the running environment for the cupcake.
After installation is complete, confirm the cupcake scripts are in your bin path: 'collapse_isoforms_by_sam.py -h', and the usage will occur on the screen, which means it works normal.
